### ISSUE TITLE
getSingleResultOrNull() method is not returning single result if the query has @OneToMany relation

### ISSUE THREAD

**Issue by nivinps** [hibernate/hibernate-reactive#956]
Hi,

The query has @OneToMany relationship then it is throwing an exception while using `getSingleResultOrNull()` method. The same query will works with `getSingleResult()`. I am using 1.0.0.CR9 version.

```
AcademicYearDBO dbo = sessionFactory.withSession(s->s.createQuery(" select dbo from AcademicYearDBO dbo " +
                                                      " left join fetch dbo.academicYearDetails where dbo.id=96", AcademicYearDBO.class).getSingleResult()).await().indefinitely(); //this will works
System.out.println(dbo.getAcademicYearDetails().size()); // for example here it will print size as 24

 AcademicYearDBO dbo1 = sessionFactory.withSession(s->s.createQuery(" select dbo from AcademicYearDBO dbo " +
                " left join fetch dbo.academicYearDetails where dbo.id=96", AcademicYearDBO.class).getSingleResultOrNull()).await().indefinitely(); // here it is throwing an exception
```

```
@Getter
@Setter
@Entity
@Table(name = "erp_academic_year")
public class AcademicYearDBO implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "erp_academic_year_id")
    private Integer id;

    @Column(name = "academic_year")
    private Integer academicYear;

    @Column(name = "academic_year_name")
    private String academicYearName;

    @Column(name = "record_status")
    private char recordStatus;

    @Column(name = "is_current_academic_year")
    private Boolean isCurrentAcademicYear;

    @Column(name = "created_users_id")
    private Integer createdUsersId;

    @Column(name = "modified_users_id")
    private Integer modifiedUsersId;

    @OneToMany(mappedBy = "academicYearDBO",cascade = CascadeType.ALL)
    private Set<AcademicYearDetaisDBO> academicYearDetails;
}
```

```
@Getter
@Setter
@Entity
@Table(name = "erp_academic_year_detail")
public class AcademicYearDetaisDBO implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "erp_academic_year_detail_id")
    private Integer id;

    @ManyToOne
    @JoinColumn(name = "erp_campus_id")
    private CampusDBO campusDBO;

    @ManyToOne
    @JoinColumn(name = "erp_academic_year_id")
    private AcademicYearDBO academicYearDBO;

    @Column(name = "record_status")
    private char recordStatus;

    @Column(name = "created_users_id")
    private Integer createdUsersId;

    @Column(name = "modified_users_id")
    private Integer modifiedUsersId;
}
```

Exception:

```
javax.persistence.NonUniqueResultException: query did not return a unique result:  there were 24results
	at org.hibernate.reactive.session.ReactiveQuery.lambda$getReactiveSingleResultOrNull$0(ReactiveQuery.java:52) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:616) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:591) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975) ~[na:1.8.0_301]
	at com.ibm.asyncutil.iteration.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:127) ~[asyncutil-0.1.0.jar:na]
	at com.ibm.asyncutil.iteration.AsyncTrampoline$TrampolineInternal.lambda$unroll$0(AsyncTrampoline.java:123) ~[asyncutil-0.1.0.jar:na]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975) ~[na:1.8.0_301]
	at io.vertx.core.Future.lambda$toCompletionStage$2(Future.java:360) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:124) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:62) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:179) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:102) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:35) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.core.Promise.complete(Promise.java:66) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.Promise.handle(Promise.java:51) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.Promise.handle(Promise.java:29) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:124) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:62) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:179) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:279) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:98) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$1(NetSocketImpl.java:97) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:240) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:130) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.NetSocketImpl.lambda$handleMessage$9(NetSocketImpl.java:390) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:49) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:275) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:389) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:155) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:154) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[utility.lib-2.5.3.jar:na]
	at io.vertx.mysqlclient.impl.codec.MySQLEncoder.lambda$write$0(MySQLEncoder.java:58) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.vertx.mysqlclient.impl.codec.QueryCommandBaseCodec.handleAllResultsetDecodingCompleted(QueryCommandBaseCodec.java:173) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.vertx.mysqlclient.impl.codec.ExtendedQueryCommandBaseCodec.handleAllResultsetDecodingCompleted(ExtendedQueryCommandBaseCodec.java:51) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.vertx.mysqlclient.impl.codec.QueryCommandBaseCodec.handleSingleResultsetDecodingCompleted(QueryCommandBaseCodec.java:136) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.vertx.mysqlclient.impl.codec.QueryCommandBaseCodec.handleRows(QueryCommandBaseCodec.java:124) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.vertx.mysqlclient.impl.codec.QueryCommandBaseCodec.decodePayload(QueryCommandBaseCodec.java:67) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.vertx.mysqlclient.impl.codec.ExtendedQueryCommandCodec.decodePayload(ExtendedQueryCommandCodec.java:109) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.vertx.mysqlclient.impl.codec.MySQLDecoder.decodePacket(MySQLDecoder.java:77) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.vertx.mysqlclient.impl.codec.MySQLDecoder.decode(MySQLDecoder.java:66) ~[vertx-mysql-client-4.1.2.jar:4.1.2]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507) ~[utility.lib-2.5.3.jar:na]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446) ~[utility.lib-2.5.3.jar:na]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581) ~[utility.lib-2.5.3.jar:na]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) ~[utility.lib-2.5.3.jar:na]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[utility.lib-2.5.3.jar:na]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[utility.lib-2.5.3.jar:na]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[utility.lib-2.5.3.jar:na]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_301]
```

**Comment by gavinking** [comment]
I've fixed this bug in #957.

In the meantime, a workaround is to use a `distinct` query.

**Comment by nivinps** [comment]
> In the meantime, a workaround is to use a distinct query.

Sure. Thanks @gavinking
